### PR TITLE
feat: create CDD component CodeEditor using the code documentation

### PR DESCRIPTION
**Pull request by BelenQuinteros68**
Storybook URL: https://ditmar.github.io/leetcode-spa/feature-leetcode-253/?path=/story/component-catalog-codeeditor--default

### CODE REVIEW

File: src/component-catalog/CodeEditor/CodeEditor.constants.ts [hunk 1]
```diff
@@ -0,0 +1,21 @@
+import type { Language } from './CodeEditor.types';
```
Comment (Ditmar) [inline src/component-catalog/CodeEditor/CodeEditor.constants.ts:1]: The constants file is well-structured, but consider adding comments to clarify the purpose of each constant for better maintainability.

**Suggestion:**
```typescript
Add comments explaining the purpose of `SUPPORTED_LANGUAGES` and `DEFAULT_CODE_TEMPLATES`.
```

File: src/component-catalog/CodeEditor/CodeEditor.hook.ts [hunk 2]
```diff
@@ -0,0 +1,113 @@
...
+    setIsOutputOpen,
+    handleLanguageChange,
+    handleRun,
+    handleSubmit,
+    handleReset,
+    increaseFontSize,
+    decreaseFontSize,
+    toggleEditorTheme,
+  };
+};
```
Comment (Ditmar) [inline src/component-catalog/CodeEditor/CodeEditor.hook.ts:214]: The hook implementation is clear, but ensure that all state updates are properly handled to avoid potential race conditions.

**Suggestion:**
```typescript
Consider using a reducer if the state management becomes more complex in the future.
```

File: src/component-catalog/CodeEditor/CodeEditor.test.tsx [hunk 3]
```diff
@@ -0,0 +1,98 @@
+import { fireEvent, render, screen, waitFor } from '@testing-library/react';
```
Comment (Ditmar) [inline src/component-catalog/CodeEditor/CodeEditor.test.tsx:1]: The test file is a good start, but it lacks coverage for edge cases and error handling scenarios.

**Suggestion:**
```typescript
Add tests for error handling and edge cases to ensure robustness.
```

**Review by Ditmar — CHANGES_REQUESTED** [review]
The implementation of the `CodeEditor` component follows the CDD principles and integrates various functionalities as specified in the ticket. However, there are some missing requirements related to unit tests and documentation that need to be addressed. Additionally, the use of `any` type is not present, and the TypeScript types are well-defined. Overall, the implementation is solid but requires some adjustments to fully meet the acceptance criteria.

### 🎫 Ticket Fulfillment (85/100)

**✅ Implemented Requirements:**
- Implement the CodeEditor component using Material UI layout primitives
- Integrate a code editing area using Monaco Editor or CodeMirror
- Create a language selector with supported languages (Python, JavaScript, TypeScript, Java, C++)
- Add a toolbar with Run, Submit, and Reset buttons
- Implement loading/executing state on Run/Submit with disabled buttons and CircularProgress
- Display output panel with test case results, runtime, memory usage, and error messages
- Show test result items with pass/fail indicators and diff for failed cases
- Add a resizable splitter between editor and output on desktop
- Implement a stacked layout for mobile with collapsible output panel
- Include font size controls and theme toggle in the toolbar
- Add keyboard shortcuts for running and submitting code
- Ensure full responsiveness and test at specified breakpoints
- Define TypeScript types in CodeEditor.types.ts
- Use styled() or sx for styles in CodeEditor.styles.ts
- Write MDX documentation in English
- Ensure accessibility compliance with axe / ARIA checks

**❌ Missing Requirements:**
- Create unit tests for various functionalities
- Develop Storybook stories for different states


### ⚠️ Detected Side Effects

- The addition of new dependencies in package.json may affect the overall bundle size and loading performance.

---

**Conclusion:** The implementation is largely compliant with the ticket requirements, but missing tests and documentation need to be addressed before approval.

File: src/component-catalog/CodeEditor/CodeEditor.tsx [hunk 4]
```diff
@@ -0,0 +1,471 @@
...
+    handleSubmit,
+    handleReset,
+  } = useCodeEditor(props);
+
+  const renderEditor = () => (
+    <Suspense fallback={<Skeleton variant="rectangular" height="100%" />}>
+      <MonacoEditor
+        height="100%"
+        language={language}
+        theme="vs-dark"
```
Comment (luisGustaJaitaGutierrez) [inline src/component-catalog/CodeEditor/CodeEditor.tsx:132]: **Severity:** Medium
**File:** `src/component-catalog/CodeEditor/CodeEditor.tsx`
###  Problem
- The toolbar currently lacks visible controls for font size and theme.
- `MonacoEditor` is configured with a fixed `theme="vs-dark"` and `fontSize: 14`.
- This is a small feature gap relative to the ticket expectations.

###  Reason
The component already has a working toolbar and editor integration.
Adding light/dark and font-size controls is a small enhancement that improves usability without restructuring the component.

### Suggested solution
- Add two small buttons for `+` / `−` font size.
- Add one theme toggle button.
- Keep these controls adjacent to the existing language selector.

File: src/component-catalog/CodeEditor/CodeEditor.tsx [hunk 5]
```diff
@@ -0,0 +1,471 @@
...
+          sx={{
+            height: isMobile ? '42%' : '34%',
+            minHeight: 0,
+            display: 'flex',
+            flexDirection: 'column',
+          }}
+        >
+          {isMobile ? (
+            <Collapse
+              in={true}
```
Comment (luisGustaJaitaGutierrez) [inline src/component-catalog/CodeEditor/CodeEditor.tsx]: **Severity:** Medium
**File:** `src/component-catalog/CodeEditor/CodeEditor.tsx`
**Change Type:** BUGFIX

### Problem
- The `Collapse` property in mobile mode is hardcoded with `in={true}`.

- The hook already exposes `isOutputOpen`, but the component never uses it.

### Reason
The mobile layout is already stacked and mostly correct.

This is a small fix that restores the intended collapse behavior without affecting the desktop experience.

### Suggested Solution
- Replace `in={true}` with `in={isOutputOpen}`.

- Add a simple toggle button or icon that changes `setIsOutputOpen`.

- Use an accessible tag like `aria-expanded`.

File: src/component-catalog/CodeEditor/CodeEditor.tsx [hunk 6]
```diff
@@ -0,0 +1,471 @@
...
+
+import { SUPPORTED_LANGUAGES } from './CodeEditor.constants';
+import { useCodeEditor } from './CodeEditor.hook';
+import {
+  CodeEditorLayout,
+  CodeEditorRoot,
+  EditorArea,
+  OutputPanel,
+  OutputToolbar,
+  ResizeHandle,
```
Comment (luisGustaJaitaGutierrez) [inline src/component-catalog/CodeEditor/CodeEditor.tsx:31]: **Severity:** Medium
**File:** `src/component-catalog/CodeEditor/CodeEditor.tsx`
**Exchange Type:** FEATURE

### Problem
- `ResizeHandle` is a styled `Box` with hover styles but no dragging behavior.
- The code visually represents a splitter, but it does not provide the resizable interaction the ticket requests.

### Reason
The existing layout is correct for desktop, and the issue is only interaction.
It can be fixed by replacing the handle with a small resizable utility or adding pointer drag logic.

### Suggested solution
- If a local `Resizable` component exists, integrate it.
- Otherwise, keep the current `ResizeHandle` and add a small drag handler around it.
- Ensure the editor/output split still collapses gracefully on mobile.

### Minimal impact
- Maintain the current editor/output layout.
- Add just enough behavior to make the splitter functional.
- No large refactor of the desktop stack.

File: src/component-catalog/CodeEditor/CodeEditor.tsx [hunk 7]
```diff
@@ -0,0 +1,471 @@
...
+
+            <Typography color="grey.100" fontFamily="monospace">
+              Expected: [1,2]
+            </Typography>
+          </TestResultItem>
+        </>
+      )}
+
+      {!isExecuting && executionResult?.errorMessage && (
+        <Alert severity="error">{executionResult.errorMessage}</Alert>
```
Comment (luisGustaJaitaGutierrez) [inline src/component-catalog/CodeEditor/CodeEditor.tsx:191]: **Severity:** Medium
**File:** `src/component-catalog/CodeEditor/CodeEditor.tsx`
**Change Type:** FEATURE

### Problem
- `executionResult.runtimeMs` and `executionResult.memoryMb` exist in the simulated types and results, but are never displayed.

- The output panel currently only shows test results and errors.

### Reason
The component already handles `executionResult` correctly.

--

File: src/component-catalog/CodeEditor/CodeEditor.docs.mdx [hunk 8]
```diff
@@ -0,0 +1,194 @@
...
+
+- Full editor workspace
+- Dedicated output panel
+- Visual separator between sections
+
+### Mobile
+
+- Stacked layout
+- Optimized spacing
+- Collapsible output section
```
Comment (luisGustaJaitaGutierrez) [inline src/component-catalog/CodeEditor/CodeEditor.docs.mdx]: **Severity:** Low
**File:** `src/component-catalog/CodeEditor/CodeEditor.docs.mdx`
**Keyline:** 102
**Change Type:** DOCUMENTATION

### Issue
- The documentation describes a collapsible behavior on mobile devices that is not currently implemented.

- Specifically, the mobile output section is described as collapsible on line 102.

### Reason
The current component is already well-aligned with the ticket in most areas.

This is a documentation fix to maintain Storybook accuracy and avoid overpromising.

File: src/component-catalog/CodeEditor/CodeEditor.test.tsx [hunk 9]
```diff
@@ -0,0 +1,141 @@
...
+        }}
+      />
+    );
+
+    expect(screen.getByText(/✗ Test case 1/i)).toBeInTheDocument();
+    expect(screen.getByText(/Expected:/i)).toBeInTheDocument();
+    expect(screen.getByText(/Received:/i)).toBeInTheDocument();
+  });
+
+  it('handles run without onRun callback', async () => {
```
Comment (luisGustaJaitaGutierrez) [inline src/component-catalog/CodeEditor/CodeEditor.test.tsx:114]: **Severity:** Low
**File:** `src/component-catalog/CodeEditor/CodeEditor.test.tsx`
**Key Line:** 112
**Change Type:** TEST COVERAGE

###  Issue

* Existing tests cover rendering, language changes, callback invocation, and execution result display.
* They do not verify that the Run, Submit, and Reset actions are disabled while execution is in progress.

### 📌Reason

The component already supports an `isExecuting` state and disables the main action buttons accordingly.

Adding a focused test for this behavior would improve confidence in the UI state management without requiring any component changes.

### Suggested Fix

* Add a test that verifies the Run, Submit, and Reset buttons are disabled when `isExecuting` is `true`.
* Keep the test focused on existing behavior already implemented by the component.

###  Minimal Impact

* No component changes required.
* Only a small addition to the test file is needed.

**Review by luisGustaJaitaGutierrez — COMMENTED** [review]
The review was conducted without creating conflicts with previous comments. Small changes are requested for improved functionality.

File: src/component-catalog/CodeEditor/CodeEditor.hook.ts [hunk 10]
```diff
@@ -0,0 +1,214 @@
...
+
+  const handleLanguageChange = useCallback((nextLanguage: Language) => {
+    dispatch({
+      type: 'CHANGE_LANGUAGE',
+      language: nextLanguage,
+    });
+  }, []);
+
+  const setCode = useCallback((code: string) => {
+    dispatch({
```
Comment (Ditmar) [inline src/component-catalog/CodeEditor/CodeEditor.hook.ts:112]: The error handling in the `handleRun` and `handleSubmit` functions catches errors but does not provide a way to inform the user of the failure beyond setting the execution result. This could lead to a poor user experience if the error is not displayed in the UI.

**Suggestion:**
```typescript
Consider adding a user-facing error message or notification when an error occurs during execution.
```

File: src/component-catalog/CodeEditor/CodeEditor.hook.ts [hunk 11]
```diff
@@ -0,0 +1,214 @@
...
+          errorMessage: error instanceof Error ? error.message : 'Unexpected submit error',
+          tests: [],
+        },
+      });
+    }
+  }, [onSubmit, state.code, state.language]);
+
+  const handleReset = useCallback(() => {
+    dispatch({
+      type: 'RESET',
```
Comment (Ditmar) [inline src/component-catalog/CodeEditor/CodeEditor.hook.ts:171]: The `handleReset` function resets the editor state but does not clear the execution result. This could lead to confusion if the user expects a fresh start.

**Suggestion:**
```typescript
Ensure that the execution result is also cleared when resetting the editor.
```

File: src/component-catalog/CodeEditor/CodeEditor.tsx [hunk 12]
```diff
@@ -0,0 +1,471 @@
...
+          <TestResultItem>
+            <Typography color="common.white" fontWeight={500}>
+              Case 2
+            </Typography>
+
+            <Typography color="grey.100" fontFamily="monospace">
+              Input: [3,2,4], 6
+            </Typography>
+
+            <Typography color="grey.100" fontFamily="monospace">
```
Comment (Ditmar) [inline src/component-catalog/CodeEditor/CodeEditor.tsx:183]: The `renderOutput` function handles rendering of execution results but does not account for cases where the execution result might be undefined or in an unexpected format.

**Suggestion:**
```typescript
Add checks to ensure that the execution result is valid before attempting to render it.
```

**Review by Ditmar — CHANGES_REQUESTED** [review]
The PR introduces a significant new feature with the CodeEditor component, including support for multiple programming languages and execution capabilities. The code appears well-structured and follows good practices. However, there are some areas for improvement regarding error handling and maintainability, particularly in the way execution results are managed.
---

**Conclusion:** While the new features are promising, the identified issues with error handling and state management need to be addressed to ensure a robust user experience.

File: src/component-catalog/CodeEditor/CodeEditor.hook.ts (hunk 10 above)
Comment (BelenQuinteros68) [reply src/component-catalog/CodeEditor/CodeEditor.hook.ts:112]: The current implementation already provides user-facing feedback when an error occurs in "handleRun" or "handleSubmit". Errors are stored in "executionResult.errorMessage" and rendered in the UI using a Material UI "Alert" component within the output panel, allowing users to immediately see execution or submission failures.

File: src/component-catalog/CodeEditor/CodeEditor.hook.ts (hunk 11 above)
Comment (BelenQuinteros68) [reply src/component-catalog/CodeEditor/CodeEditor.hook.ts:171]: The current implementation already provides user-facing feedback when an error occurs in "handleRun" or "handleSubmit". Errors are stored in "executionResult.errorMessage" and rendered in the UI using a Material UI "Alert" component within the output panel, allowing users to immediately see execution or submission failures.

File: src/component-catalog/CodeEditor/CodeEditor.tsx (hunk 4 above)
Comment (BelenQuinteros68) [reply src/component-catalog/CodeEditor/CodeEditor.tsx:132]: I reviewed the suggestion and also discussed it with the instructor to confirm the expected scope of the component. I was advised to keep the implementation aligned with the approved Figma design and not introduce additional font size or theme controls in this iteration.

For that reason, I kept the editor configuration fixed to maintain consistency with the visual specification defined for the component.

**Review by luisGustaJaitaGutierrez — APPROVED** [review]
You addressed the comments correctly in the review; it's noticeable that you corrected them differently than indicated, but you did correct them, therefore I approve your PR
